modules/_labs/core/react/lib/theming/applyDirection.ts

```diff
@@ -0,0 +1,47 @@
+import * as React from 'react';
+import {ContentDirection} from './types';
+
+/**
+ * Recursively set direction on all child components
+ */
+function recursiveMap(
+  direction: ContentDirection | undefined,
+  children: React.ReactNode
+): React.ReactElement<any> | React.ReactNode {
+  return React.Children.map(children, child => {
+    if (!React.isValidElement(child)) {
+      return child;
+    }
+
+    if (child.props.children) {
+      return React.cloneElement(child, {
+        children: applyDirection(direction, child.props.children),
+        direction: direction,
+      });
+    }
+
+    return React.cloneElement(child as React.ReactElement, {
+      direction: direction,
+    });
+  });
+}
+
+/**
+ * This is used to apply `dir={direction}` to the top level element
+ */
+function applyDirection(
```

Comment:
@stephanerangaya This addresses the issue you were running into (where the `direction` prop was not being passed to all styled components from the parent component.

I'm unsure about the performance implications on this, but it may be worth it for the better devX.

On one hand, we could do it manually:
- It's a huge pain to pass `direction={direction}` to every single styled component. This is just asking for developer error.
- We have to manually ensure `dir={direction}` is added to the top level container of every component
- This is the fastest performance wise

On the other hand, we could use `applyDirection` function:
- This barely adds any extra markup in components
- Automatically adds `dir` attribute and passes `direction` to all children, reducing the likelihood of dev errors
- It's iterating over every child of a component and creating a clone (double the memory) on render. Not sure how quick this is and what garbage collection is like on the clones, but it isn't ideal. Would love some input from anyone who has more experience with the perf impacts of this.

Reply:
The comment doesn't say _why_ this function is required. I'm unsure as well. Aren't we defining `direction` on the theme object which is provided through React context? Why do we need to add `dir` and `direction` to every child component? Doesn't the `styled` helper provide it? Is it for projected content we don't control?

Reply:
Yes, the `styled` helper providers the direction from the context. However, we also want to support a prop so users can change the direction explicitly for specific components. When they pass the prop to the component, we then need to share that with all of the styled components. That's what this is achieving.

Alternatively, @stephanerangaya and I have been talking about requiring the user to wrap the component with another `CanvasProvider` with `direction` changed if they need to swap the direction for a specific component.

Reply:
@anicholls Do we still need this if we're wrapping components in another CanvasProvider?

Reply:
@stephanerangaya If we want to use `CanvasProvider` to locally override the direction, then we need to change our logic for setting the body `dir` attribute. Right now, the last rendered `CanvasProvider` will take precedence, so trying to override locally would set the attribute on the entire page.

Reply:
After discussing it with @stephanerangaya, we decided to just use a nested `CanvasProvider` to change the direction of a nested component. Since we recently made the change to only set the `dir` attribute for the children of a `CanvasProvider`, we no longer need to worry about overriding the body `dir` (we're going to have consumers control this themselves). 

This may be a bit more perf overhead than using a `direction` prop, but it's not a very common occurrence. Also, it avoids the high likelihood of human error when passing `direction` to all styled components, and alternatively, avoids using an unconventional render wrapper like the `applyDirection` function I wrote. Since we added a no-op for `InputProvider` in #272, we shouldn't need to worry about nested providers.

Lastly, this method enforces the fact that direction is just part of the theme object, so the API for setting a local theme is the same as setting a local content direction.

Example:
```tsx
<CanvasProvider theme={createCanvasTheme({direction: ContentDirection.RTL})}>

  // A bunch of components

  <CanvasProvider theme={createCanvasTheme({direction: ContentDirection.LTR})}>
    <Slider /> // Sliders should always be LTR
  </CanvasProvider>

  // Some more components

</CanvasProvider>
```

Reply:
👍